src/Compilers/Core/Portable/Collections/Rope.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// A representation of a string of characters that requires O(1) extra space to concatenate two ropes.
+    /// </summary>
+    internal abstract class Rope
+    {
+        public static readonly Rope Empty = ForString("");
+        public abstract override string ToString();
+        public abstract int Length { get; }
+        protected abstract IEnumerable<char> GetChars();
+
+        /// <summary>
+        /// A rope can wrap a simple string.
+        /// </summary>
+        public static Rope ForString(string s)
+        {
+            if (s == null)
+                throw new ArgumentNullException(nameof(s));
+
+            return new StringRope(s);
+        }
+
+        /// <summary>
+        /// A rope can be formed from the concatenation of two ropes.
+        /// </summary>
+        public static Rope Concat(Rope r1, Rope r2)
+        {
+            if (r1 == null)
+                throw new ArgumentNullException(nameof(r1));
+
+            if (r2 == null)
+                throw new ArgumentNullException(nameof(r2));
+
+            return
+                r1.Length == 0 ? r2 :
+                r2.Length == 0 ? r1 :
+                checked(r1.Length + r2.Length < 32) ? ForString(r1.ToString() + r2.ToString()) :
+                new ConcatRope(r1, r2);
+        }
+
+        /// <summary>
+        /// Two ropes are "the same" if they represent the same sequence of characters.
+        /// </summary>
+        public override bool Equals(object obj)
+        {
+            if (!(obj is Rope other) || Length != other.Length)
+                return false;
+            if (Length == 0)
+                return true;
+            var chars0 = GetChars().GetEnumerator();
+            var chars1 = other.GetChars().GetEnumerator();
+            while (chars0.MoveNext() && chars1.MoveNext())
+            {
+                if (chars0.Current != chars1.Current)
+                    return false;
+            }
+
+            return true;
+        }
+        public override int GetHashCode()
+        {
+            int result = Length;
+            foreach (char c in GetChars())
+                result = Hash.Combine((int)c, result);
```

Comment:
note: there are ways you can design ropes such that the sub parts contain their hashes, and they can be combined such that any equivalent ropes would have the same hash (i did this in my Rope implementation at Google :)).

If this is an operation called a lot, that may be worth doing.  However, if the compiler has no need to hash these, then that work isn't necessary.

Reply:
The compiler hashes these, but rarely.  The only times it does is when a string appears as a case in a switch.

---
In reply to: [313170138](https://github.com/dotnet/roslyn/pull/37915#discussion_r313170138) [](ancestors = 313170138)

Reply:
Removing most of those now.

---
In reply to: [313902881](https://github.com/dotnet/roslyn/pull/37915#discussion_r313902881) [](ancestors = 313902881,313170138)